### ISSUE TITLE
TMC on 5.2 generates unexpected lambda for Melange

### ISSUE THREAD

**Issue by EduardoRFS** [ocaml/ocaml#12989]
After #12958 a new kind of lambda is being generated by the OCaml compiler, which triggers an issue in the Melange(@anmonteiro) compiler.

After debugging, this seems like an issue in actual OCaml as the following lambda is unexpected, it is essentially expecting an exit(Lstaticraise) to return to write the pointer, but by definition an exit never returns, making the `setfield_ptr` deadcode.

```clojure
(setfield_ptr(heap-init)_computed dst/282 offset/281
  (exit 1 0))
```

## Troubleshooting

This happens because the `Choice.lambda` at https://github.com/ocaml/ocaml/blob/trunk/lambda/tmc.ml#L641 always wraps all the lambdas, including the ones that are never going to return.

## OCaml Code

This is the original OCaml code that generates the lambda in 5.2
```ocaml
let[@tail_mod_cons] rec input_lines ic =
  match () with () -> () :: input_lines ic | exception End_of_file -> []
```

**Comment by lthls** [comment]
In OCaml, it's fine. Even though this `setfield` is never executed, the handler for `exit` also does `setfield` so in the end everything is fine.
(This happens to break the invariant that all exit handlers are in tail position with respect to their handlers, which `Simplif.simplify_exits` relies on, but the try-with disables the optimisation so we're safe for now.)
Do you have a pointer to the issue in Melange ? It might be easier to patch Melange than to make sure the compiler never generates this.

**Comment by EduardoRFS** [comment]
Antonio did a fix for Melange, but it seems like this should not be a valid lambda anyway, while it is low priority, my bet it is that this will bite OCaml in the future.

**Comment by lthls** [comment]
Why wouldn't this be valid ? It's similar to `1 + raise Not_found`, the `1 +` part is dead code but it still works as intended.

**Comment by gasche** [comment]
Out of curiosity, can you give references to the Melange-side issue and fix?

**Comment by gasche** [comment]
Note: the example

```ocaml
let rec input_lines ic =
  match () with () -> () :: input_lines ic | exception End_of_file -> []
```

generates the following lambda code *without* the TMC transformation:

```
(setglobal Test!
  (letrec
    (input_lines/270
       (function ic/271
         (catch
           (try (exit 1 0) with exn/274
             (if (== exn/274 (field_imm 11 (global Stdlib!))) 0
               (reraise exn/274)))
          with (1 val/273[int])
           (makeblock 0 (int,*) 0 (apply input_lines/270 ic/271)))))
    (makeblock 0 input_lines/270)))
```

The `(setfield_ptr ... (exit 1 0))` discussed here comes from the TMC transformation of `try (exit 1 0) with ...`. That code looks a bit odd, but it is explained by the following comment in translcore.ml:

https://github.com/ocaml/ocaml/blob/b2cd7e59e5b563b6a0527014689b7224bb678661/lambda/translcore.ml#L1096-L1112

**Comment by EduardoRFS** [comment]
@lthls right, it should work, but this is an Lstaticraise, so in general that's exactly one of the goals of simplif right? I know that it is semantically valid, but my gut reaction was that this invariant being broken is undesirable.

@gasche the following code, but in general is just an invariant that the Melange compiler relies on, I personally think this is a bad design and it should definitely handle such cases.

https://github.com/melange-re/melange/blob/080055227b3198643056280736cfba0b5e73d5bc/jscomp/core/lam_compile.ml#L1631

**Comment by lthls** [comment]
> @lthls right, it should work, but this is an Lstaticraise, so in general that's exactly one of the goals of simplif right? I know that it is semantically valid, but my gut reaction was that this invariant being broken is undesirable.

Simplif performs optimizations, but as far as I know it was never meant to normalize its input. In addition, there are several individual passes, and TMC runs as the last one, so no other pass will be able to simplify the output of TMC.

**Comment by gasche** [comment]
I thought a bit about how to change the TMC pass to avoid this `setfield ... (exit ...)` pattern. I think that the solution would be to specialize `exit` in the transformation, or slightly more generally to add new logic to transform non-returning expressions. This is a bit subtle and some work (maybe not a lot), so I am not eager to do it for now. If someone wants to give it a try, they should feel free, but we also have to be mindful of not introducing too much complexity there.

I tend to agree with @lthls' position that the generated code is valid and reasonable, especially in the context where the `exit` we are talking about was already not in tail position. It would probably be better for Melange to clarify its own assumptions about `exit` than to have us bend backward to accommodate a semi-broken model. (But: I don't know how easy/hard this is to fix in Melange, and I recognize that it suffers from an even worse manpower shortage than we do. Apologies.). I will accordingly close this as "not planned". This is not intended to prevent people from working on this if they really want to.
